modules/roles/server/controllers/roles.controller.js

```diff
@@ -228,145 +308,159 @@ function roleModule(logger, userModule) {
+    })
+    .catch(error =>
+    {
+      return sendServerError(res, error);
+    });
 
-    return deferred.promise
-      .then((data) => {
-        res.status(data.code).send(data.data);
-        }, (error) => {
-        res.status(error.code).send(error.error);
-      });
   }
 
-  // --------------------------- Private Function Definitions ----------------------------
-
-   /*
-   * Recursive function to get the subroles of a role
-   *
-   * @param {parentRoleName}  The role name to find subroles for
-   * @param {data} The list of roles currently in the database
-   * @param {subroles} the list of subroles
+  /*
+   * Returns a tree object representing the role tree
+   * begining at the target role or 'admin' role if no
+   * target is specified
    */
-   function getRolesByParent(parentRoleName, data, subroles)
-   {
-      var directDescendants = _.filter(data, 'parent', parentRoleName);
+  function getRoleTree(req, res, next)
+  {
+    rootRole = req.params.id ? req.params.id : 'admin';
+    getAllRoles()
+    .then((data) =>
+    {
+      var roleTree = buildTree(rootRole, data);
+      return res.status(200).send(roleTree);
+    })
+    .catch((error) =>
+    {
+      return sendServerError(res,"could not get role tree");
+    });
+  }
 
-      if(directDescendants && directDescendants.length > 0)
-      {
-        _.forEach(directDescendants, function(descendant)
-        {
-          subroles.push(descendant._id);
 
-          getRolesByParent(descendant._id, data, subroles);
-        })
-      }
+  /*
+   * Returns a formatted role tree object starting from the target
+   */
+  function buildTree(targetId, data)
+  {
+    var tree = {
+      _id: targetId,
+      children: getDirectDescendants(targetId, data)
+    };
+    for (var child in tree.children)
+    {
+      tree.children[child] = buildTree(tree.children[child]._id, data);
+    }
+    return tree;
+  }
 
-      return subroles; 
-   }
 
-   function roleExists(roleName)
-   {
-    var deferred = q.defer();
-    var roleCount = Roles.count({_id: roleName}, (err, count) =>{
+  /*
+   *  Returns a promise to fetch all roles in the role collection
+   */
+  function getAllRoles()
+  {
+    return Roles.find({}).exec();
+  }
 
-      logger.info(count);
-      if(err)
-      {
-        return deferred.resolve(false);
-      }
-      return deferred.resolve(count > 0);
+  /*
+   * This function returns a list of direct descendants for a given role
+   */
+  function getDirectDescendants(roleId, allRoles)
+  {
+    var results = _.filter(allRoles, function(val)
+    {
+      return val._doc.parent == roleId;
     });
+    var formatted = results.map((x) =>
+    {
+      var obj = {
+        _id: x._id,
+        children: []
+      };
+      return obj;
+    });
+    return formatted;
+  }
 
-    return deferred.promise;
+  function sendServerError(res, error)
+  {
+    logger.error(error);
+    res.status(500).send(
```

Comment:
There were many situations where this function was called but a 500 status code would be inaccurate.  This should be a passed in argument instead.